src/rust/ide/view/graph-editor/src/component/node.rs

```diff
@@ -570,34 +576,78 @@ impl Node {
             out.source.skip   <+ action_bar.action_skip;
             out.source.freeze <+ action_bar.action_freeze;
             eval out.hover ((t) action_bar.set_visibility(t));
+       }
 
 
-            // === Errors on Node ===
+        // === Visualizations & Errors ===
 
-            is_error_set <- frp.error.map(|err| err.is_some());
+        let hover_onset_delay = DelayedAnimation::new(network);
+        hover_onset_delay.set_delay(VIS_PREVIEW_ONSET_MS);
+        hover_onset_delay.set_duration(0.0);
 
-            frp.source.error <+ frp.set_error.map(f!([model](error) {
-                model.set_error(error.as_ref());
-                error.clone()
-            }));
+        frp::extend! { network
 
+            frp.source.error <+ frp.set_error;
+            is_error_set <- frp.error.map(|err| err.is_some());
+            no_error_set <- not(&is_error_set);
             error_color_anim.target <+ frp.error.map(f!([style](error)
                 Self::error_color(error,&style))
             );
 
-            eval error_color_anim.value ((value) model.set_error_color(value));
-
-
-            // === Visualization ===
-
             eval frp.set_visualization ((t) model.visualization.frp.set_visualization.emit(t));
             visualization_enabled_frp <- bool(&frp.disable_visualization,&frp.enable_visualization);
             eval visualization_enabled_frp ((enabled) model.action_bar.set_action_visibility_state(enabled));
-            no_error_set          <- not(&is_error_set);
+
+            // Show preview visualisation after some delay, depending on whether we show an error
+            // or are in quick preview mode. Also, omit the delay if we don't have a tooltip to show.
+            has_tooltip <- model.output.frp.tooltip.map(|tt| tt.has_content());
+            preview_show_delay <- all3(&frp.quick_preview_vis,&is_error_set,&has_tooltip).map(
+                |(quick_preview,is_error,has_tooltip)| {
+                if !has_tooltip {
+                    return 0.0;
```

Comment:
I'd be thankful if we do not use in the code early-return statements if not needed necessarily. The functional, dataflow program structure is more obvious to the reader.